### ISSUE TITLE
Tuning TabNet Hyperparameters K-Fold CV

### ISSUE THREAD

**Issue by FernandoAMDM** [jrzaurin/pytorch-widedeep#45]
In Scikit-learn a common workflow would be to define your hyperparameter space, the model to be tested and fit like that:

```
model_cv = RandomizedSearchCV(model(), hyperparameter_space, cv=cv, n_iter=n_iter)


model_cv.fit(x_train, y_train)
```

Now in pytorch-widedeep the recommended path for training a single model is:

```
trainer = Trainer(model, objective="binary", metrics=[Accuracy])


trainer.fit(X_wide=X_wide, X_tab=X_tab, target=target, n_epochs=5, batch_size=256, val_split=0.1)
```

So, how could I do a random search for the best hyperparameter of a model (for example: TabNet or TabTransformer) in a cross validated way ? If I just put the trainer object in the RandomizedSearchCV of the first example and define the hyperparameter space of the model could I make it work right ? If that's not possible, what would be other way to do it ?

**Comment by jrzaurin** [comment]
Hey @FernandoAMDM, thanks for the question!

Well, my first comment would be to not use `RandomizedSearchCV` or `GridSearchCV` these days, when you have superior Bayesian methods (`Hyperopt`, `Optuna` or even `Ax` for Pytorch). However, if you do need or want to use those methods you would have to code your own class to be passed to `RandomizedSearchCV`.  Something like this:

```python
class MyModel:
    def __init__(self, *model_args, **trainer_args, **training_args):
        ...

    def fit(self, X, y):
       self.model = WideDeep(deeptabular=TabNet(**self.model_args))        
       self.trainer = Trainer(self.model, **self.trainer_args)
       self.trainer.fit(X_tab=X_tab, target=target, **self.training_args)

    def predict(self, X):
       self.trainer.predict(X)
```

now, bear in mind the following. Models like `TabNet` and `TabTransfomer` take forever to run. As soon as the `hyperparameter_space` is of some size, on a relatively big dataset, if you do CV with just 3 folds, this is going to take forever. 

My suggestion would be to do a standard train/valid/test split, do hyperparameter optimisation in the valid set and get test metrics in the test set. 

Anyway, we have a slack channel: https://join.slack.com/t/pytorch-widedeep/shared_invite/zt-soss7stf-iXpVuLeKZz8lGTnxxtHtTw

If you want to join we can discuss further and maybe I can pass you functioning code

Let me know if this helps!

**Comment by FernandoAMDM** [comment]
Thank you, I will give it a try

**Comment by jrzaurin** [comment]
Closing it for now. Let me know if you have any other Question
